Feature/hydroscat

davidbenncsiro opened the pull request:
Hi @doizuc

It has taken awhile for me to get the HydroScat code to a point where I'm relatively happy with it.

Please review the code. Note the addition of dependency information re: HydroScat in the README.

 I've tested this with an emulator (in aquasense package; see README) and with a HydroScat-6 instrument.

I have some questions for your consideration that I will add as individual comments below.

I look forward to your review comments.

davidbenncsiro commented:
What is the procedure for adding a JSON config file fragment for a new instrument?

As you can see from conflicts, I made changes to the config file, but they may not be how you want things done. I'd appreciate your guidance.

davidbenncsiro commented:
I have modified `run()` in the `Instrument` class in order to allow commands to be sent even when no serial input. Is this OK? Is there a better way? Could it conflict with other instruments, e.g. causing an additional data packet to be requested per `run()` call in LISST?

Again, your guidance is sought.

davidbenncsiro commented:
What is the best way to handle a clean flush and close of I/O when setup is saved while in running state? In the normal course of events, you wouldn't expect someone to save setup during a run, but it's not prevented. Is there a way of (preferably generically) disabling the setup dialog/button during a run?

davidbenncsiro commented:
Can you point to example code successfully using checkboxes in Inlinino? I'm using them in the setup dialog for HydroScat but not seeing the values change.

davidbenncsiro commented:
Along the way I added code for handling spinner widgets (in `gui.py`). Leave in or remove? My setup dialog used spinners at first but then I moved to text boxes. There's obviously little point in retaining code that won't be used.

davidbenncsiro commented:
Note also the commented code in a few places (marked with TODO) to avoid having to install dependencies for other instruments. That will obviously be uncommented before a merge to `dev`.

doizuc commented:
Hi @davidbenncsiro,
Thank you for adding the HydroScat to Inlinino. I will review your code and answer your questions asap.
Best,
Nils

doizuc commented:
> What is the procedure for adding a JSON config file fragment for a new instrument?

I keep the JSON config file in the repository as an example of configuration for various instruments. When the file is touched by Inlinino, the json library compresses it in one line. I use my text editor to automatically reformat the configuration file before I commit it to keep it human-readable and track modifications.


> I have modified run() in the Instrument class in order to allow commands to be sent even when no serial input. Is this OK? Is there a better way? Could it conflict with other instruments, e.g. causing an additional data packet to be requested per run() call in LISST?

The Apogee and Ontrak instruments have similar issues. The solution you propose is great. I suggest moving the statement `self.write_to_interface()` outside the `if data: ... else: ...` statement and removing the `self.write_to_interface()` in the `handle_packet` method.This way it applies in either case (with or without data received).  I will rework on the LISST class and change the implementation in Apogee to use the generic `Instrument.run()` method instead of overwriting it.


> What is the best way to handle a clean flush and close of I/O when setup is saved while in running state? In the normal course of events, you wouldn't expect someone to save setup during a run, but it's not prevented. Is there a way of (preferably generically) disabling the setup dialog/button during a run?

I agree that during the normal course of operations the user is not expected to save setups while running an instrument. However, we found it useful for some instruments when first setting them up to be able to go back and force between running the instrument and setting it up. When the setup is saved by the user, the instrument interface is closed with `HydroScat.close()` before changes are applied. User has to manually re-open the connection after, and the instrument is restarted with `HydroScat.open()`. Hence, I would use these methods or `init` methods called within this methods to flush I/O. In short, a user clicking on close instrument and then open instrument is the same as a user clicking on "setup>save" and open instrument.


> Can you point to example code successfully using checkboxes in Inlinino? I'm using them in the setup dialog for HydroScat but not seeing the values change.

The code handling the setup modal has been through a number of iteration and could certainly be cleaned sorry for the confusion there. The Ontrak instruments use checkbox. I don't distribute the ontrak module with Inlinino source code but you can download it [here](https://www.ontrak.net/images/adu_python_dll.zip).


> Along the way I added code for handling spinner widgets (in gui.py). Leave in or remove? My setup dialog used spinners at first but then I moved to text boxes. There's obviously little point in retaining code that won't be used.

The Ontrak instrument is using spinbox, however it uses the keyword `spinbox` instead of `sb`. Your implementation is more generic and consistent with other code, so I will clean up Ontrak there. It's fine to keep it.


> Note also the commented code in a few places (marked with TODO) to avoid having to install dependencies for other instruments. That will obviously be uncommented before a merge to dev.

Noted. `ImportError` could be used as other might encounter the same issue with the OnTrak module.

davidbenncsiro commented:
@doizuc I've made the changes to `Instrument.run()` as suggested:
* removed `write_to_interface()` from `handle_packet()` and
* moved outside of `if data:` in `run()`
  * I moved it to be near the end, after `if data:`, just before `except InterfaceException as e:`    
  * let me know if you are happy with that; I've tested it for HydroScat.
 
Do you think there's any other code changes required?
* `apogee.py` could presumably use write_to_interface() in future now if desired.
* `lisst.py` calls it but now it's unguarded in `run()`, that means it will always send a `GX` packet whenever it's called. Is that OK?

davidbenncsiro commented:
Re: `HydroScat.close()` and flushing,looking again, I don't think there's anything that needs to be / can be flushed, so may need nothing more there.

davidbenncsiro commented:
Re:
> Noted. ImportError could be used as other might encounter the same issue with the OnTrak module.

are you happy with the following?
~~~
from inlinino.instruments.nmea import NMEA
try:
    from inlinino.instruments.ontrak import Ontrak, USBADUHIDInterface
except ImportError:
    pass
from inlinino.instruments.satlantic import Satlantic
~~~

davidbenncsiro commented:
In [inlinino/widgets/hypernav/calibrate_dialog.py](https://github.com/OceanOptics/Inlinino/pull/11/files#diff-c4036281cf3598ffda912feebe298578e4579bffd3382fe743990ef0e5918386) we have:
~~~
try:
    from hypernav.calibrate import calibration_report
    from hypernav.io import HyperNav as HyperNavIO
except IndexError:
    HyperNavIO = None
~~~
Should this be `ImportError`?

doizuc commented:
> In [inlinino/widgets/hypernav/calibrate_dialog.py](https://github.com/OceanOptics/Inlinino/pull/11/files#diff-c4036281cf3598ffda912feebe298578e4579bffd3382fe743990ef0e5918386) we have:
> 
> ```
> try:
>     from hypernav.calibrate import calibration_report
>     from hypernav.io import HyperNav as HyperNavIO
> except IndexError:
>     HyperNavIO = None
> ```
> 
> Should this be `ImportError`?

Good catch. It was also recently corrected in dev branch.

doizuc commented:
> @doizuc I've made the changes to `Instrument.run()` as suggested:
> 
> * removed `write_to_interface()` from `handle_packet()` and
> * moved outside of `if data:` in `run()`
>   
>   * I moved it to be near the end, after `if data:`, just before `except InterfaceException as e:`
>   * let me know if you are happy with that; I've tested it for HydroScat.
> 
> As you mentioned, you'll be able to rework Apogee and LISST since `apogee.py` could presumably use `write_to_interface()` now if desired. `lisst.py` overrides it but now it's unguarded in `run()`, that means it will always send a `GX` packet whenever it's called. Is that OK?

Correct, I will adjust Apogee and LISST when I merge the PR in dev.

doizuc commented:
> Re:
> 
> > Noted. ImportError could be used as other might encounter the same issue with the OnTrak module.
> 
> are you happy with the following or can we do something better than just `pass` here?
> 
> ```
> from inlinino.instruments.nmea import NMEA
> try:
>     from inlinino.instruments.ontrak import Ontrak, USBADUHIDInterface
> except ImportError:
>     pass
> from inlinino.instruments.satlantic import Satlantic
> ```
> 
> and
> 
> ```
> if platform.system() == 'Windows':
>     try:
>         from inlinino.resources.ontrak import aduhid
>     except ImportError:
>         pass
> else:
>     aduhid = None
> ```

The `try: ... except ImportError: ...` statement in inlinino/gui.py is not needed, as the one in the ontrack.py file should suffice. For the second one in inlinino/instruments/ontrack.py, could you replace the `pass` by aduhid = None. There is one more error that needs to be handled under Windows, while you're at it if you can add line 484 after `raise InterfaceException(...)` the exception below.

            if aduhid is None:
                raise InterfaceException('USB-ADUHID driver not found.')

doizuc commented on inlinino/gui.py:
```diff
@@ -1290,8 +1306,11 @@ def start(self, instrument_index=None):
-                                    'lisst': LISST, 'nmea': NMEA, 'ontrak': Ontrak, 'satlantic': Satlantic,
-                                    'sunav1': SunaV1, 'sunav2': SunaV2, 'taratsg': TaraTSG}
+                                    'lisst': LISST, 'nmea': NMEA,
+                                    #'ontrak': Ontrak,
```
Ontrak can now be uncommented here, following discussion on handling ImportError in ontrak specific to Windows.

davidbenncsiro replied:
Missed this one. Thanks.

doizuc commented on inlinino/instruments/hydroscat.py:
```diff
@@ -0,0 +1,278 @@
+            self.change_state("READY")
+
+
+    def log_start(self):
```
Typically in Inlinino, instruments start sampling as soon as the connection with the instrument is open, when the user click on the 'Serial Port: Open' button. However, data is recorded only when the user start on 'Log Data: Start' the main reason is to make sure the instrument is measuring properly (e.g. no bubbles) before we record data to simplify data processing later on. 

For consistency with other instruments and if I understood correctly the code, I wonder if instead of overloading the `log_start` and `log_stop` methods, it wouldn't be more appropriate to overload the `open` and `close` methods.

davidbenncsiro replied:
I like it! I puzzled over this one for awhile. I'm no longer overriding `log_start|stop()` or `open()` since hitting the `Open` button calls `init()` which now immediately sends a `START` command to HydroScat so that sampling begins straight away, as you suggest (with Start/Stop buttons just controlling logging which I had not fully appreciated). I also simplified the state machine in the process, removing the `READY` state. So, the only method I'm overriding now is `close()`.

Let me know what you think.

davidbenncsiro commented:
> > Re:
> > > Noted. ImportError could be used as other might encounter the same issue with the OnTrak module.
> > 
> > 
> > are you happy with the following or can we do something better than just `pass` here?
> > ```
> > from inlinino.instruments.nmea import NMEA
> > try:
> >     from inlinino.instruments.ontrak import Ontrak, USBADUHIDInterface
> > except ImportError:
> >     pass
> > from inlinino.instruments.satlantic import Satlantic
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```
> > if platform.system() == 'Windows':
> >     try:
> >         from inlinino.resources.ontrak import aduhid
> >     except ImportError:
> >         pass
> > else:
> >     aduhid = None
> > ```
> 
> The `try: ... except ImportError: ...` statement in inlinino/gui.py is not needed, as the one in the ontrack.py file should suffice. For the second one in inlinino/instruments/ontrack.py, could you replace the `pass` by aduhid = None. There is one more error that needs to be handled under Windows, while you're at it if you can add line 484 after `raise InterfaceException(...)` the exception below.
> 
> ```
>         if aduhid is None:
>             raise InterfaceException('USB-ADUHID driver not found.')
> ```

I've removed the `ImportError` catch in `gui.py` and replaced `pass` by `aduhid = None`.

Which file is line 484 in? `inlinino/instruments/ontrack.py` only has 467 lines.

doizuc commented:
> > > Re:
> > > > Noted. ImportError could be used as other might encounter the same issue with the OnTrak module.
> > > 
> > > 
> > > are you happy with the following or can we do something better than just `pass` here?
> > > ```
> > > from inlinino.instruments.nmea import NMEA
> > > try:
> > >     from inlinino.instruments.ontrak import Ontrak, USBADUHIDInterface
> > > except ImportError:
> > >     pass
> > > from inlinino.instruments.satlantic import Satlantic
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and
> > > ```
> > > if platform.system() == 'Windows':
> > >     try:
> > >         from inlinino.resources.ontrak import aduhid
> > >     except ImportError:
> > >         pass
> > > else:
> > >     aduhid = None
> > > ```
> > 
> > 
> > The `try: ... except ImportError: ...` statement in inlinino/gui.py is not needed, as the one in the ontrack.py file should suffice. For the second one in inlinino/instruments/ontrack.py, could you replace the `pass` by aduhid = None. There is one more error that needs to be handled under Windows, while you're at it if you can add line 484 after `raise InterfaceException(...)` the exception below.
> > ```
> >         if aduhid is None:
> >             raise InterfaceException('USB-ADUHID driver not found.')
> > ```
> 
> I've removed the `ImportError` catch in `gui.py` and replaced `pass` by `aduhid = None`.
> 
> Which file is line 484 in? `inlinino/instruments/ontrack.py` only has 467 lines.

I looked at `dev` branch... in this branch it's line 433.

doizuc commented:
Is there a specific version of aquasense I should use to test your code? I just tried with the latest and have a miss-match between the arguments Inlinino was passing to the aquasense package in `aquasense.hydroscat.HydroScat`.

Very cool the HydroScat emulator for Arduino! I don't have an Arduino handy right now... do you have an emulator that runs on a computer that I could pipe to a virtual serial port?

davidbenncsiro commented:
> Very cool the HydroScat emulator for Arduino! I don't have an Arduino handy right now... do you have an emulator that runs on a computer that I could pipe to a virtual serial port?

Sadly not at the moment. I have considered it though. Something that would be good to add. If I recall correctly you have some (BB9 and other?) emulators for Arduino in the Inlinino repo and note also some for virtual serial port (https://github.com/OceanOptics/Inlinino/blob/master/test/mock_serial_sensor.py) I should adapt the HydroScat one to a virtual serial port in my aquasense repo.

doizuc commented:
Correct, I did write an emulator outputting data frames from several instruments including BB9 [Simulino.cpp](https://github.com/OceanOptics/Inlinino/blob/tb-app/arduino/Simulino.cpp), I somehow removed it from the main branch. It doesn't handle any input.  I will wait to have an Arduino handy, no worries.

davidbenncsiro commented:
> Correct, I did write an emulator outputting data frames from several instruments including BB9 [Simulino.cpp](https://github.com/OceanOptics/Inlinino/blob/tb-app/arduino/Simulino.cpp), I somehow removed it from the main branch. It doesn't handle any input. I will wait to have an Arduino handy, no worries.

I will however take your virtual serial port suggestion on board and add that to `aquasense` when I get the chance. It will be another nice system level test tool.

davidbenncsiro commented:
> > Can you point to example code successfully using checkboxes in Inlinino? I'm using them in the setup dialog for HydroScat but not seeing the values change.
> 
> The code handling the setup modal has been through a number of iteration and could certainly be cleaned sorry for the confusion t[here](https://www.ontrak.net/images/adu_python_dll.zip). The Ontrak instruments use checkbox. I don't distribute the ontrak module with Inlinino source code but you can download it here.

I have this working now with a couple extra lines of code and an addition to one line.

I'm using `cb_` as checkbox prefix. Would you prefer `checkbox_` to disambiguate from `combobox_`? If so, I'll make that change tomorrow. It's getting on for midnight here (South Australia) and I have one or two things to do before getting some shut-eye. :)

doizuc commented:
`checkbox_` seems better to avoid confusion as you pointed out.

I just went over the code and ran it (without instrument), it looks good to me. I slightly adjusted setup_hydroscat.ui, adding layouts for the window to scale properly when shrank or expanded, I will commit it to this branch if ok with you?

With respect to the code I had one minor bug line 218 of inlinino/gui.py due to an empty `self.instrument.spectrum_plot_x_values` array, I corrected as follow:

                    min_x = min(min_x, min(x)) if min_x is not None else (min(x) if len(x) > 0 else 0)
                    max_x = max(max_x, max(x)) if max_x is not None else (max(x) if len(x) > 0 else 1)

davidbenncsiro commented:
> `checkbox_` seems better to avoid confusion as you pointed out.
> 
> I just went over the code and ran it (without instrument), it looks good to me. I slightly adjusted setup_hydroscat.ui, adding layouts for the window to scale properly when shrank or expanded, I will commit it to this branch if ok with you?
> 
> With respect to the code I had one minor bug line 218 of inlinino/gui.py due to an empty `self.instrument.spectrum_plot_x_values` array, I corrected as follow:
> 
> ```
>                 min_x = min(min_x, min(x)) if min_x is not None else (min(x) if len(x) > 0 else 0)
>                 max_x = max(max_x, max(x)) if max_x is not None else (max(x) if len(x) > 0 else 1)
> ```

Please go ahead and commit changes. Thanks Nils! I will change `cb_` to `checkbox_` soon.

davidbenncsiro commented:
The only other question I have right now is that when I start getting data from the instrument I currently dump calibration info out as INFO log entries. The person I'm collaborating expressed interest in that info being added to one of the output files, e.g. the CSV file. Is there are point in the code where you think that would make sense?

This is where I currently do it: https://github.com/OceanOptics/Inlinino/blob/feature/hydroscat/inlinino/instruments/hydroscat.py#L161

doizuc commented:
To dump additional information at the head of the logged data, I suggest overloading the Logger you use (accessible under self._log_raw or self._log_prod) as it's done for Satlantic instruments. 

[inlinino/instruments/satlantic.py](https://github.com/OceanOptics/Inlinino/blob/cb7b07b6d279aedec2b35b303632e10d9e54391a/inlinino/instruments/satlantic.py#L363)

davidbenncsiro commented:
> To dump additional information at the head of the logged data, I suggest overloading the Logger you use (accessible under self._log_raw or self._log_prod) as it's done for Satlantic instruments.
> 
> [inlinino/instruments/satlantic.py](https://github.com/OceanOptics/Inlinino/blob/cb7b07b6d279aedec2b35b303632e10d9e54391a/inlinino/instruments/satlantic.py#L363)

Rather than rushing into changing this, I'm inclined to go with what is currently there, talk with my collaborator in the New Year and see whether she wants it changed and if so, which log to write to (raw of prod). Once I know I can create a new issue with a more targeted feature branch. Do you have thoughts about that?

doizuc commented:
I touched up the setup modal, fixed a bug preventing the addition of new HydroScat instrument (e.g. manufacturer, model, and serial number were missing), added error checking on the path to calibration file (I was able to feed an incorrect path and had no errors, leading to no data displayed after), fixed bug preventing update of spectrum plot if no channels are selected to display on timeseries plot. Details are in my past commits. Arduino simulator of HydroScat-6 worked great. Let me know how it runs on your side, if all good, I'm happy to merge the PR.

davidbenncsiro commented on inlinino/instruments/hydroscat.py:
```diff
@@ -28,15 +29,14 @@ class HydroScat(Instrument):
+    def __init__(self, uuid, cfg, *args, **kwargs):
+        super().__init__(uuid, cfg, *args, setup=False, **kwargs)
         # Instrument state machine
+        self.hydroscat: aquasense.hydroscat.HydroScat = None
```
I'm curious about this line. Not familiar with this syntax. The denotation appears to be something like:
~~~
aquasense.hydroscat.HydroScat = None if self.hydroscat
~~~
but I suspect that's not it.

Should this assignment to `None` be in the `try...import hydroscat...except` block?

doizuc replied:
This syntax is used for type hints in python (c.f. [typing](https://docs.python.org/3/library/typing.html)), it's not used by the Python runtime but can be used by your IDE to make suggestions and check variable types for you. The syntax is as follow:

    variable_name: variable_type = variable_value

In this case, it's equivalent to `self.hydroscat = None`. Omitting `:aquasense.hydroscat.HydroScat` won't change how the Python runtime execute the code.

I did set `self.hydroscat` in the __init__ function, as a personal preference and for consistency with the rest of Inlinino's code base (all class variables are defined in `__init__`), the code would still run fine without setting up the variable. The value is changed during initialization as `self.setup()` is called during `__init__`.

davidbenncsiro replied:
@doizuc, I was obviously not thinking at all clearly when I looked at that line of code.

In fact I use and advocate for type hints in Python all the time in my code (e.g. see https://bitbucket.csiro.au/projects/AQRS/repos/aquasense/browse/src/aquasense/hydroscat/extract_raw.py).

I can only assume that my confusion arose from its use in an instance variable since I have probably only used it in function/method signatures so far.

Also, I completely agree with you about initialising all instance variables. I am sometimes less strict about this in Python than in other languages like C++, C# or Java, for no very good reason, so thanks for picking me up on that.

davidbenncsiro reviewed (commented):
Thanks for the improvements/fixes @doizuc. I added a comment/question re: line 35 but that's about all I can think of.

davidbenncsiro commented:
It seems good to me via the emulator @doizuc.

The only thing I have seen in the past and still see is, when data is being received, displayed, logged, if I open the Setup dialog and hit the Save button, I'll often see this kind of error:
~~~
Exception in thread HydroScat HS080339:
Traceback (most recent call last):
  File "C:\Users\ben29w\Anaconda3\lib\threading.py", line 926, in _bootstrap_inner
    self.run()
  File "C:\Users\ben29w\Anaconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\ben29w\Projects\working_copy\AquaticSensors\Inlinino\inlinino\instruments\__init__.py", line 241, in run
    self.close(wait_thread_join=False)
  File "C:\Users\ben29w\Projects\working_copy\AquaticSensors\Inlinino\inlinino\instruments\hydroscat.py", line 179, in close
    super().close(wait_thread_join)
  File "C:\Users\ben29w\Projects\working_copy\AquaticSensors\Inlinino\inlinino\instruments\__init__.py", line 181, in close
    self._interface.stop()
  File "C:\Users\ben29w\Projects\working_copy\AquaticSensors\Inlinino\inlinino\instruments\__init__.py", line 438, in stop
    self._serial.cancel_read()
  File "C:\Users\ben29w\Anaconda3\lib\site-packages\serial\serialwin32.py", line 465, in cancel_read
    self._cancel_overlapped_io(self._overlapped_read)
  File "C:\Users\ben29w\Anaconda3\lib\site-packages\serial\serialwin32.py", line 456, in _cancel_overlapped_io
    ctypes.byref(overlapped),
TypeError: byref() argument must be a ctypes instance, not 'NoneType
~~~
Is that something worth addressing? If so, create a separate issue for this or address it here?

Thanks.

doizuc commented:
#8d1377c1 will fix the issue you were having. The serial interface was destroyed before being closed...
